### PR TITLE
Warn of invalid travel plans pre-takeoff, pulse invalid travel routes

### PR DESCRIPTION
**Pull request by tehhowch**
This PR introduces a warning upon attempting to take off from a planet when the player's selected "Travel Plan" or "Travel Destination" is no longer possible (much like if you don't have enough cargo space to leave without selling outfits in cargo). If the player continues launching, the travel plan / destination is cleared as appropriate.

The player can view any invalid portion of their travel plan in the standard map view that shows your intended route. The invalid segments are drawn [alternating between transparent and their normal color](https://youtu.be/0WSBg4FfSKk) to indicate that these segments are of special interest to the player. If the chosen planet in the final system to land upon is no longer accessible, the final system of the plan will be drawn with a red ring around it that also pulses.

This PR prevents the player from
1) Having a `player.TravelDestination` that corresponds to an inaccessible planet (which has caused application crashes in the past)
2) Having a travel plan that routes through restricted wormholes without the access key
3) Having a travel plan that needs a jump-capable drive, but only having a link-following drive

(It also makes sure the player is notified of any carried ships that will be sold even if a mission failure would also happen, though this can/should get cherry-picked to `master` anyway.)

### CODE REVIEW

**Comment by jafdy** [comment]
Could no longer having a jump drive be a travel restriction for point 1 as well.

**Comment by tehhowch** [comment]
Absolutely it could (and totally forgot about that case 👍 ) I'll add it when I get a chance

**Comment by tehhowch** [comment]
I extended the check to include jump drive removal, and also moved it into a PlayerInfo function for use by other classes. This allows a warning before taking off, preventing the player from unwittingly departing, tapping Jump, and only then discovering they are in the system they were pointed at and not the one their old travel plan specified.
image

**Comment by Bladewood** [comment]
That's pretty vague.

**Comment by tehhowch** [comment]
It'll also never pop up unless you do one of the following after already setting up a travel plan:
 - swap a Jump Drive for a Hyperdrive (if your plan moves from an system to a non-linked neighbor)
- route through a wormhole that has required attributes and then remove those attributes by 1) uninstalling the needed outfit, or 2) change your flagship for one without the outfit
- set a destination planet that has required attributes and then 1) uninstall the needed outfit, or 2) swap your flagship for one without the outfit.

I'm willing to bet that any player who receives the message can remember that they just changed their flagship or its installed outfits.

That said, I'm open to a better message. You can see the current TakeOffIfReady warning messages here: [PlanetPanel.cpp Line 254](https://github.com/endless-sky/endless-sky/blob/master/source/PlanetPanel.cpp#L254).

**Comment by endless-sky** [comment]
The landing messages can be fixed much more simply. (See 72cea97.)

**Comment by tehhowch** [comment]
@endless-sky - 72cea97 does not restore landing messages. This PR does. I will update the check (because that check is nicer) but your commit is missing additional requirements to display the default landing messages.

File: source/PlayerInfo.cpp [hunk 1]
```diff
@@ -1728,6 +1735,49 @@ bool PlayerInfo::HasTravelPlan() const
 
 
 
+// Verify the player's travel plan and destination are still valid, in case the flagship was changed and/or
+// any needed outfits were removed.
+bool PlayerInfo::HasInvalidTravelPlan() const
+{
+	if(travelDestination && !travelDestination->IsAccessible(flagship.get()))
+		return true;
+	if(!travelPlan.empty())
```
Comment (endless-sky) [inline source/PlayerInfo.cpp]: Since this never gets called in performance-critical code, it'd make more sense to just avoid duplicating all the code below by instead just checking:

```C++
if(!travelPlan.empty() && !DistanceMap(*this).HasRoute(travelPlan.front())
    return true;
```

The only downside would be that it would then pop up the warning message if the plan was invalidated due to systems becoming "unexplored" (e.g. the Pug invasion). But, I wonder whether showing that message is really useful or necessary, anyway.

Comment (tehhowch) [inline source/PlayerInfo.cpp]: I don't think that correctly detects when the player has two routes to a location, but the faster (e.g. automatic and most likely current) route is through restricted space. Example: Going from Zosma[Memory] to Cardea, as long as you have a Jump Drive, DistanceMap knows you can (eventually) get there, but if you had a Quantum Keystone when you made the route, your current map is set to use the wormhole in Terminus.

I wrote the inspection just in case the player had defined some custom route (Shift clicking), as that also handles the case where the player sets some custom route that is different that the automatic route.

For the Jump Drive -> Hyperdrive swap, the "you can't make this jump" error is already raised, but only *after* the player jumps through however many preceeding jumps along hyperspace lanes. HasInvalidTravelPlan will catch that immediately (and !DistanceMap.HasRoute may fail to catch it if there are hyperspace lanes along some other path).

**Comment by tehhowch** [comment]
Per @endless-sky 's [comment](https://github.com/endless-sky/endless-sky/pull/2601#discussion_r120253169), I made a branch that uses DistanceMap.HasRoute() instead of inspecting the planned route. As I identified previously, this does not fix the issues at hand, while this PR does.

To verify my claim, [checkout that branch](https://github.com/tehhowch/endless-sky/tree/TestingHasRoute), and use save file "George Costanza" below. Simply load the save and launch. If DistanceMap::HasRoute could perform accurate travel plan validation, it would prompt a warning dialog before you could take off. If you then proceed to jump, the original travel plan that was set when using the other ship is followed, rather than the route to the destination that DistanceMap found. 
Note that the game should not programmatically reset the travel plan upon takeoff without first changing how travel plans are constructed and stored, because the player can freely construct a route that is not the same as what DistanceMap might create.



[George Costanza.txt](https://github.com/endless-sky/endless-sky/files/1079138/George.Costanza.txt)
A travel plan from Memory to Aventine was specified using the Dreadnought, which has the Quantum Key Stone outfit. The flagship was then changed to the Solifuge.
<details><summary>DistanceMap:FindRoute (with Keystone)</summary>

travel_via_keystone
</details><details><summary>DistanceMap:FindRoute (no Keystone):</summary>

travel_without_keystone
</details>

**Comment by tehhowch** [comment]
With the merge of #2639 I still feel this is a good enhancement to the player experience, preventing them from embarking on otherwise fruitless travel unknowingly. 
The warning message should pop up for any currently-known "roadblocks", but (since they are not generated until after the player takes off and the date increments) does not address changes to the map in general that happen after the player has last disembarked from a planet or station. For players with Jump Drives, most link removals will not impact the ability to travel to that system, with the exception of links that are longer than the neighbor distance. For those cases, and the cases where the player only has link-following capability, the game already displays a message "You cannot jump to the selected system".

**Comment by endless-sky** [comment]
One thing that could be confusing about this is that if the player's travel plan is invalid, the invalid portion simply isn't displayed in the map. So, I'm worried that players will look at the map, see no travel plan (or a valid travel plan), then try to depart, get the warning, check the map again, and not understand what's wrong with the travel plan that's visible there.

**Comment by tehhowch** [comment]
A couple updates:
1. Removed the invalid destination planet check from HasInvalidTravelPlan()
2. Always remove invalid destination planets (this prevents auto-landing and guards against auto-selection of stellarobjects in systems that have no currently-landable objects).
3. Alter the structure of HasInvalidTravelPlan to determine which of the steps in the travel plan is the invalid one, and return its index.
4. Changed the function name to InvalidTravelPlanIndex to reflect its new purpose
5. Write a boolean wrapper to maintain the original functionality
6. Add a medium-rate pulsing color to MapPanel::DrawTravelPlan, to highlight all steps past the invalid step. I chose white since it is not used for the other travel plan route parameters. 
7. Clean up the comments used
image

**Comment by tehhowch** [comment]
While adding the pulsing link segments for invalid travel plans, I couldn't help but notice some places that could be improved. One of the fixes ensures that the arrows are always drawn on top of wormhole links - on master this wormhole direction indicator gets written over by the Terminus link:
`master`: 
image
`a02392b`:
image

I'll try to record the pulsing to get some feedback on it.

**Comment by KiLEdEnNis** [comment]
I think the pulse is way too strong
first the pulse animation should play and then reverse, instead of "off > on, off > on" it should be "off > on > off > on" so it looks smooth
and the black/white pulse seems too much of a color difference, it should pulse between less extreme colors

**Comment by tehhowch** [comment]
I didn't want to get too elaborate in the formula used for picking the color, given that it i expect it to be a rare occurrence that players depart with an invalid plan. This limits the available behaviors, since the controlling `step` value increments by one.

I can lower the alpha channel, which should reduce the strength of the color. I can also add a floor to the color rather than zero. I wanted to stay away from green, yellow, and red, as those colors are already in use for drawing the travel plan, but I could also add in another call to LineShader to blend the invalid identifier with the normal color.

**Comment by tehhowch** [comment]
pulsing from low-grey to white: https://youtu.be/t7SBVCYf0OE
alternating from strong red to white: https://youtu.be/Xg09Xaatkgo

the low-grey to white is added in 83cd891

**Comment by tehhowch** [comment]
Updated the warning:
image

**Comment by tehhowch** [comment]
Per https://github.com/endless-sky/endless-sky/pull/3138#issuecomment-338092795, only wormholes with planet descriptions are drawn.
I will let @endless-sky choose which base game wormholes should be drawn (or add some very basic descriptions to those that are indicated if requested).

(note that your travel plan will draw your route, even if it uses a non-drawn wormhole link.)

**Comment by tehhowch** [comment]
Updated base game wormholes with `description ""` and made selecting a wormhole not show the planet description panel

**Comment by tehhowch** [comment]
If #3379 merges this will contain only the relevant "invalid travel" additions.

**Comment by tehhowch** [comment]
Cleaned up the commit history

**Comment by endless-sky** [comment]
I'm still not sure I see the point of this. The current behavior is:
* If the player makes a ship change that invalidates part of their travel plan, that part of the plan simply "disappears."
* If they undo the change, the travel plan is restored.

So, the main change you're making here is to show a warning message to the player instead of just silently invalidating part of the travel plan. That's only an improvement if the number of times that the player would actually want to see and respond to that warning outnumber the times when the player would be fine with the path being invalidated automatically, and I'm not sure that's the case.

**Comment by tehhowch** [comment]
My goal here is to reduce or eliminate any frustration associated with setting a plan, and at some point while following it, making changes to the flagship that make it unable to be completed as written, but not disclosing this to the player until they are at the now-unable-to-complete step in the plan:
- Without this PR, one could capture a new ship along the way to a destination, make that ship the new flagship, and then continue traveling, only to find out that -- oops -- the necessary outfit is now on an escort and not the flagship. For the base game this is really only due to Quantum Keystones and Jump Drive/Hyperdrive, but mods can certainly have a lot more gateway outfits.
 - With this PR, that flagship swap will generate a departure notification that they forgot to check something, and viewing their route will show what part is now not working (which will help the player discern what is making it invalid).

**Comment by tehhowch** [comment]
In my opinion that is met by this pull - the only 2 ways to get an invalid plan are:

1. The player changes their flagship or outfits on their flagship (e.g. removing a keystone or jump drive)
2. A GameEvent modifies the systems or wormholes on the travel plan.

In the first (the by-far most common occurrence), the player can absolutely resolve the issue. They have yet to depart the planet, and as such, any outfits they may have sold remain available in the outfitter without requiring a save reload to before takeoff. Additionally, they may not be able to land at the invalid node, so displaying the warning will let them pick the proper flagship if needed, without detouring further.

For 2, the warning serves as a notification that the galaxy has changed significantly, and the player should probably reevaluate their plans.

I do want to revisit this pull and see if it can be done differently or more easily, so don't merge it just yet :)

**Comment by tehhowch** [comment]
The route highlight looks like this: https://youtu.be/thfxicv03jc

The pre-takeoff dialog is now properly a standalone warning, which if okayed then defers to the standard cargo checks for mission auto-failure and auto-sold cargo/drones/fighters. Previously, only one check was possible at a time (with mission failure having priority).

File: data/map.txt [hunk 2]
```diff
@@ -27298,4 +27298,3 @@ planet Zug
 	tribute 1200
 		threshold 3000
 		fleet "Large Militia" 18
-
```
Comment (Amazinite) [inline data/map.txt]: K.

**Review by Amazinite — DISMISSED** [review]
Seems to be working now.

**Review by Amazinite — COMMENTED** [review]
I tested the following scenarios while landed:
* Player was able to land on a planet but is no longer able due to lacking an attribute requirement.
* Player was able to land on a wormhole but is no longer able due to lacking an attribute.
* Player was able to pass through a wormhole but is no longer able due to lacking an attribute.
* Player was able to jump between two systems but is no longer able because their jump range changed.

I had expected behavior in the first three scenarios, that being that the now invalid section of the travel path was highlighted, or the now invalid planet destination had its system highlighted, and all of the above scenarios resulted in a warning dialog when attempting to depart. But the last scenario of changing jump range did something different than I had expected.

Here is a travel path using a higher jump range than normal.
image
After reducing my jump range (or by removing my ability to jump, both result in this behavior), I would have expected the travel path to still be visible on the map, just with a flashing path, and perhaps colored red to show that the jump is impossible (instead of green or yellow, which are only for when the jump is possible for the flagship). Instead, the system has a flashing red ring around it, which I would expect to be indicative of an invalid planet destination, not an invalid system to jump to.
image

Not saying that this behavior is wrong, just that it's not what I expected. If this is intended behavior though, no need to look into it.

I then tested the following scenarios while in flight:
* An event unlinks two systems on the player's travel path.
* An event moves two unlinked systems on the player's travel path further apart than their jump range.
* An event adds a restriction to a planet that previously didn't have one.
* An event adds a restriction to a wormhole that previously didn't have one.

I got the following behavior from the above tests:
* I got expected behavior from a planet or wormhole that became inaccessible in flight, i.e. the system with the planet or wormhole I am attempting to land on is highlighted, or the wormhole path I am attempting to follow is highlighted.
* When two systems were unlinked, what I would expect to occur is what happened when I changed my jump range while landed. That is, the now inaccessible system has a red ring around it. But when the two systems on my path were unlinked, the only change was that my travel path was cut short. The now inaccessible system was not highlighted as expected.
* Having a system on my path be moved out of range resulted in the same behavior as two systems being unlinked, i.e. I would expect the inaccessible system to be highlighted but it isn't. In either case, reaching the end of the valid path does result in an error sound and a message about the selected system being inaccessible.

Might it be worth giving the player a dialog once they enter the map with an invalid path saying "An event has made your travel path invalid," or something to that effect?

<hr />

TL;DR
* Question: Is it intended behavior that a jump that is made invalid by now being out of range or no longer able to use a jump drive puts a red ring around the system that can no longer be jumped to instead of having the travel path remain but now be shown as invalid?
* Issue: Travel paths that become invalid while in flight due to a system no longer being reachable via jumping don't have the same behavior as when landed; the red ring around the invalid system doesn't appear.
* Optional: Warn the player when they enter the map while in flight if their travel path has become invalid.

**Comment by petervdmeer** [comment]
I did a brief test of this PR;
- The pulsing route didn't show after switching ships (the invalid part of the route was just not shown at all).
- Amazinites remark about the jump range doesn't seem to have been handled yet.

**Comment by Amazinite** [comment]
Closing all PRs older than 2020 which don't have any active discussion. If we wish to revisit these PRs then they may be reopened.
